src/client/ws/jsonrpc_transport.rs

```diff
@@ -0,0 +1,94 @@
+//! JSONRPC Transport module.
+//!
+//! Wraps the underlying WebSocket transport with specific JSONRPC details.
+
+use crate::client::ws::transport::{self, WsConnectError, WsNewDnsError};
+use crate::types::jsonrpc;
+
+/// Creates a new JSONRPC WebSocket connection, represented as a Sender and Receiver pair.
+pub async fn websocket_connection(target: impl AsRef<str>) -> Result<(Sender, Receiver), WsNewDnsError> {
+	let (sender, receiver) = transport::websocket_connection(target.as_ref()).await?;
+	Ok((Sender::new(sender), Receiver::new(receiver)))
+}
+
+pub struct Sender {
+	request_id: u64,
+	transport: crate::client::ws::transport::Sender,
+}
+
+impl Sender {
+	pub fn new(transport: crate::client::ws::transport::Sender) -> Self {
+		Self { transport, request_id: 0 }
+	}
+
+	/// Inner implementation for starting either a request or a subscription.
+	async fn start_impl(
+		&mut self,
+		method: impl Into<String>,
+		params: impl Into<jsonrpc::Params>,
+	) -> Result<u64, WsConnectError> {
+		let id = self.request_id;
+		self.request_id = id.wrapping_add(1);
+
+		let request = jsonrpc::Request::Single(jsonrpc::Call::MethodCall(jsonrpc::MethodCall {
+			jsonrpc: jsonrpc::Version::V2,
+			method: method.into(),
+			params: params.into(),
+			id: jsonrpc::Id::Num(id),
+		}));
+
+		// Note that in case of an error, we "lose" the request id (as in, it will never be
+		// used). This isn't a problem, however.
```

Comment:
It's not a problem because we always increase the ID right?

Reply:
yes, it's increased on every request and it may "wrap around" when `u64::MAX` is reached.

The comment is copy-pasted from the old code.